Sources/FoundationXML/XMLParser.swift

```diff
@@ -469,33 +461,58 @@ open class XMLParser : NSObject {
     open var externalEntityResolvingPolicy: ExternalEntityResolvingPolicy = .never
     
     open var allowedExternalEntityURLs: Set<URL>?
-    
-#if os(WASI)
-    private static var _currentParser: XMLParser?
-#endif
 
+    /// The current parser context for the current thread.
+    private class _CurrentParserContext {
+        var _stack: [XMLParser] = []
+        var _current: XMLParser? {
+            return _stack.last
+        }
+    }
+
+    #if os(WASI)
+    /// The current parser associated with the current thread. (assuming no multi-threading)
+    /// FIXME: Unify the implementation with the other platforms once we unlock `threadDictionary`
+    ///        or migrate to `FoundationEssentials._ThreadLocal`.
+    private static nonisolated(unsafe) var _currentParserContext: _CurrentParserContext?
+    #else
+    /// The current parser associated with the current thread.
+    private static var _currentParserContext: _CurrentParserContext? {
+        get {
+            return Thread.current.threadDictionary["__CurrentNSXMLParser"] as? _CurrentParserContext
+        }
+        set {
+            Thread.current.threadDictionary["__CurrentNSXMLParser"] = newValue
+        }
+    }
+    #endif
+
+    /// The current parser associated with the current thread.
     internal static func currentParser() -> XMLParser? {
-#if os(WASI)
-        return _currentParser
-#else
-        if let current = Thread.current.threadDictionary["__CurrentNSXMLParser"] {
-            return current as? XMLParser
-        } else {
-            return nil
+        if let ctx = _currentParserContext {
+            return ctx._current
         }
-#endif
+        return nil
     }
-    
-    internal static func setCurrentParser(_ parser: XMLParser?) {
-#if os(WASI)
-        _currentParser = parser
-#else
-        if let p = parser {
-            Thread.current.threadDictionary["__CurrentNSXMLParser"] = p
+
+    /// Execute the given closure with the current parser set to the given parser.
+    internal static func withCurrentParser<R>(_ parser: XMLParser, _ body: () -> R) -> R {
+        var ctx: _CurrentParserContext
+        if let current = _currentParserContext {
+            // Use the existing context if it exists
+            ctx = current
         } else {
-            Thread.current.threadDictionary.removeObject(forKey: "__CurrentNSXMLParser")
+            // Create a new context in TLS
+            ctx = _CurrentParserContext()
+            _currentParserContext = ctx
         }
-#endif
+        // Push the parser onto the stack
```

Comment:
Does this need to be a stack at all? Can't the actual call stack itself be used?

```swift
let oldValue = ctx._currentParser
ctx._currentParser = parser
defer {
  ctx._currentParser = oldValue
}
```

This may simplify your design overall.

Reply:
That totally makes sense!